Self-Consistent-Field/RHF_symmetry.py

```diff
@@ -0,0 +1,186 @@
+
+assert len(S_) == len(T_) == len(V_) == len(transformers_)
+nirrep = len(S_)
+nsopi = [n for n in nsopi if n > 0]
+
+print('\nTotal time taken for integrals: %.3f seconds.' % (time.time() - t))
+t = time.time()
+
+# Build H_core: [Szabo:1996] Eqn. 3.153, pp. 141
+H_ = [T + V for (T, V) in zip(T_, V_)]
+
+# Build the orthogonalizer A = S^(-1/2) for each irrep.
+A_ = []
+for S in S_:
+    # Application of a function to a matrix requires transforming to
+    # the diagonal basis, applying the function to the diagonal form,
+    # then backtransformation to the original basis.
+    eigval, eigvec = np.linalg.eigh(S)
+    eigval_diag = np.diag(eigval ** (-1./2))
+    A = eigvec.dot(eigval_diag).dot(eigvec.T)
+    A_.append(A)
+
+# Calculate initial core guess: [Szabo:1996] pp. 145
+C_ = []
+e_ = []
+for i in range(nirrep):
+    Hp = A_[i].dot(H_[i]).dot(A_[i])  # Eqn. 3.177
+    e, C2 = np.linalg.eigh(Hp)        # Solving Eqn. 1.178
+    C = A_[i].dot(C2)                 # Back transform, Eqn. 3.174
+    C_.append(C)
+    e_.append(e)
+
+# Initial occupations are taken from the lowest eigenvalues (energies)
+# of the guess coefficients.
+e_tot = np.concatenate(e_)
+idxsort = np.argsort(e_tot)
+so_to_irrep = np.array([i for i, nso in enumerate(nsopi) for _ in range(nso)])
+lowest_occupied = so_to_irrep[idxsort][:ndocc].tolist()
+ndoccpi = [lowest_occupied.count(i) for i in range(nirrep)]
+print('Number of occupied spin orbitals per irrep:', ndoccpi)
+
+# Form (occupied) density: [Szabo:1996] Eqn. 3.145, pp. 139
+D_ = []
+for i, indocc in enumerate(ndoccpi):
+    D = np.einsum('mi,ni->mn', C_[i][:, :indocc], C_[i][:, :indocc])
+    D_.append(D)
+
+print('\nTotal time taken for setup: %.3f seconds' % (time.time() - t))
+
+print('\nStart SCF iterations:\n')
+t = time.time()
+E = 0.0
+Enuc = mol.nuclear_repulsion_energy()
+Eold = 0.0
+Dold_ = [np.zeros_like(D) for D in D_]
+F_ = [np.zeros_like(D) for D in D_]
+
+E_ = np.array([(D * (H + H)).sum() for (D, H) in zip(D_, H_)])
+E_guess = sum(E_) + Enuc
+print('SCF Iteration %3d: Energy = %4.16f' % (0, E_guess))
+
+for SCF_ITER in range(1, maxiter + 1):
+
+    # Perform the two-electron integral contraction with the density in the AO
+    # basis.
+    D_AO = transform_sotoao(D_, transformers_)
+    J_ = transform_aotoso(np.einsum("mnls,ls->mn", I_AO, D_AO), transformers_)
+    K_ = transform_aotoso(np.einsum("mlns,ls->mn", I_AO, D_AO), transformers_)
+    F_ = [H + (2 * J) - K for H, J, K in zip(H_, J_, K_)]
+    E_ = [np.einsum("mn,mn->", D, H + F) for D, H, F in zip(D_, H_, F_)]
+
+    SCF_E = sum(E_) + Enuc
+
+    print('SCF Iteration %3d: Energy = %4.16f   dE = % 1.5E' % (SCF_ITER, SCF_E, (SCF_E - Eold)))
+    if abs(SCF_E - Eold) < E_conv:
+        break
+
+    Eold = SCF_E
+
+    for h, indocc in enumerate(ndoccpi):
+        Fp = A_[h].dot(F_[h]).dot(A_[h])
+        e, C2 = np.linalg.eigh(Fp)
+        C_[h] = A_[h].dot(C2)
+        Cocc = C_[h][:, :indocc]
```

Comment:
Could consider functionalizing the doccpi tech so that we could update the irrupts here.

Reply:
Do you think it should return C, even though only D is necessary for the iterations?